File: breadth-first-search/coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0){
            return 0;
        }

        int memoir[amount+1];

        for (int i = 1; i < amount+1; i++) {
            memoir[i] = pow(2, 16);

            for (int x: coins) {
                if (i == x) {
                    memoir[i] = 1;
                    break;
                }

                if (i-x > 0 && memoir[i] > memoir[i-x] + 1) {
                    memoir[i] = memoir[i-x] + 1;
                }
            }
        }

        return memoir[amount] == pow(2, 16) ? -1 : memoir[amount];
    }
};

File: array/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [inf] * (amount + 1)
        dp[0] = 0

        for value in range(1, amount + 1):
            for coin in coins:
                prev_value = value - coin
                if prev_value >= 0:
                    dp[value] = min(dp[value], dp[prev_value] + 1)


        return dp[amount] if dp[amount] != inf else -1